ppgan/models/generators/generator_firstorder.py

```diff
@@ -0,0 +1,256 @@
+            value_total = 0
+            for scale in self.disc_scales:
+                key = 'prediction_map_%s' % scale
+                value = ((1 - discriminator_maps_generated[key])**2).mean()
+                value_total += self.loss_weights['generator_gan'] * value
+            loss_values['gen_gan'] = value_total
+            # Feature matching Loss
+            if sum(self.loss_weights['feature_matching']) != 0:
+                value_total = 0
+                for scale in self.disc_scales:
+                    key = 'feature_maps_%s' % scale
+                    for i, (a, b) in enumerate(
+                            zip(discriminator_maps_real[key],
+                                discriminator_maps_generated[key])):
+
+                        if self.loss_weights['feature_matching'][i] == 0:
+                            continue
+                        value = paddle.abs(a - b).mean()
+                        value_total += self.loss_weights['feature_matching'][
+                            i] * value
+                loss_values['feature_matching'] = value_total
+        if (self.loss_weights['equivariance_value'] +
+                self.loss_weights['equivariance_jacobian']) != 0:
+            transform = Transform(x['driving'].shape[0],
+                                  **self.train_params['transform_params'])
+            transformed_frame = transform.transform_frame(x['driving'])
+            transformed_kp = self.kp_extractor(transformed_frame)
+            generated['transformed_frame'] = transformed_frame
+            generated['transformed_kp'] = transformed_kp
+
+            # Value loss part
+            if self.loss_weights['equivariance_value'] != 0:
+                value = paddle.abs(
+                    kp_driving['value'] -
+                    transform.warp_coordinates(transformed_kp['value'])).mean()
+                loss_values['equivariance_value'] = self.loss_weights[
+                    'equivariance_value'] * value
+
+            # jacobian loss part
+            if self.loss_weights['equivariance_jacobian'] != 0:
+                jacobian_transformed = paddle.matmul(
+                    *broadcast(transform.jacobian(transformed_kp['value']),
+                               transformed_kp['jacobian']))
+                normed_driving = paddle.inverse(kp_driving['jacobian'])
+                normed_transformed = jacobian_transformed
+                value = paddle.matmul(
+                    *broadcast(normed_driving, normed_transformed))
+                eye = paddle.tensor.eye(2, dtype='float32').reshape(
+                    (1, 1, 2, 2))
+                eye = paddle.tile(eye, [1, value.shape[1], 1, 1])
+                value = paddle.abs(eye - value).mean()
+                loss_values['equivariance_jacobian'] = self.loss_weights[
+                    'equivariance_jacobian'] * value
+        return loss_values, generated
+
+
+class Vgg19(nn.Layer):
```

Comment:
Vgg19 -> VGG19

Reply:
can this reuse criterion/perceptual_loss.py?